### DISCUSSION TITLE
Launching DPF

### DISCUSSION THREAD

**Discussion by germa89** [ansys/pydpf-core#588] [Q&A]
Hi all,

As far as I know there are 2 ways to connect to DPF: InProcess and gRPC.

To connect to gRPC, we use ``dpf.connect_to_server(ip, port)``.

What would be the case for ``InProcess``.

Furthermore, the way to connect to ``InProcess`` relies on any envvar or MAPDL directory where it looks for the binaries??

I ask because I need to implement DPF to read PyMAPDL results and I need to handle both cases, local (Inprocess,grpc) and remote (grpc). So I need to be able to tell PyMAPDL how to connect to each of one and how.

The workflow I want to achieve is:

```py
def connect_to_dfp():
    if "DPF_port" not in os.environ("DPF_PORT"):
        # Try local connection
        try:
            return dpf.connect_to_inprocess() # I need to know this function or something like
        except ConnectionError:
            DPF_port = Default_DPF_port # 50054

    else:
        DPF_port = os.environ("DPF_PORT")

    return dpf.connect_server(port=DPF_port, ip=DPF_port
```

So I need:
* The supposed ``dpf.connect_to_inprocess`` function
* (Bonus) The exception ``ConnectionError``

Pinging @rafacanton @PProfizi @GuillemBarroso @cbellot000

**Comment by rafacanton** [comment]
Hi @germa89, @GuillemBarroso. The default behavior is to connect in-process. To handle all possible connection possibilities, you can take a look at https://dpf.docs.pyansys.com/examples/00-basic/11-server_types.html. Basically, if you don't set anything, the communication will be done in-process. You can change this default with the `dpf.SERVER_CONFIGURATION` variable. Finally, you can build specific variables/operators in different server instances as shown in the Example.

FYI @PProfizi

**Reply by germa89** [reply]
Ok... Maybe I need to be a bit more specific. I'm updating the text above.

**Reply by germa89** [reply]
Btw, I did check that example. But it does no really show how the connection is set though...

**Reply by rafacanton** [reply]
@germa89, @GuillemBarroso, @PProfizi You can set up the server configuration at the beginning of the session. Any dpf entity created after setting it up will be created in that server. Take a look at this snippet:

```py
import ansys.dpf.core as dpf

def connect_to_dpf(mode, external_ip=None, external_port=None):
    if mode == "InProcess":
        dpf.server.set_server_configuration(dpf.server_factory.AvailableServerConfigs.InProcessServer)
        srvr = dpf.server.start_local_server()
    elif mode == "LocalGrpc":
        dpf.server.set_server_configuration(dpf.server_factory.AvailableServerConfigs.GrpcServer)
        srvr = dpf.server.start_local_server()
    elif mode == "RemoteGrpc":
        dpf.server.set_server_configuration(dpf.server_factory.AvailableServerConfigs.GrpcServer)
        if external_ip is not None and external_port is not None:
            srvr = dpf.server.connect_to_server(ip=external_ip, port=external_port)
        else:
            raise Exception("external_ip and external_port should be provided for RemoteGrpc communication")
    print("Server setup: "+str(srvr))


connect_to_dpf("InProcess")

f1 = dpf.fields_factory.create_3d_vector_field(3)
print("f1 Server: "+str(f1._server))

connect_to_dpf("LocalGrpc")

f2 = dpf.fields_factory.create_3d_vector_field(3)
print("f2 Server: "+str(f2._server))

connect_to_dpf("RemoteGrpc", external_ip=my_ip_string, external_port=my_port_number)

f3 = dpf.fields_factory.create_3d_vector_field(3)
print("f3 Server: "+str(f3._server))

```

This produces (after setting `my_ip_string` and `my_port_number` to appropriate values of an already running Dpf server, in this case in a Linux machine):

![image](https://user-images.githubusercontent.com/107186344/198993035-bda139a1-3576-47e3-99df-6aa0e4bedf05.png)

In the local connections you can specify `ansys_path` as well to connect to a specific local Ansys install (in case several are available).

**Reply by germa89** [reply]
This should work. Let me try to implement it in the coming days.
